dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/AuditManager.java

```diff
@@ -0,0 +1,105 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import com.google.common.primitives.Primitives;
+import org.apache.qpid.jms.JmsQueue;
+import org.apache.qpid.jms.JmsTopic;
+import org.hisp.dhis.render.RenderService;
+import org.springframework.jms.core.JmsTemplate;
+import org.springframework.stereotype.Component;
+
+import javax.jms.Destination;
+
+/**
+ * @author Morten Olav Hansen <mortenoh@gmail.com>
+ */
+@Component
+public class AuditManager
+{
+    private final JmsTemplate jmsTemplate;
+    private final RenderService renderService;
+
+    public AuditManager(
+        JmsTemplate jmsTemplate,
+        RenderService renderService
+    )
+    {
+        this.jmsTemplate = jmsTemplate;
+        this.renderService = renderService;
+    }
+
+    public void send( Destination destination, Audit audit )
+    {
+        send( destination, audit, null );
+    }
+
+    public void send( Destination destination, Audit audit, Object data )
+    {
+        String payload = null;
+
+        if ( data != null )
+        {
+            Class<?> klass = data.getClass();
+
+            if ( klass.isPrimitive() || Primitives.isWrapperType( klass ) )
+            {
+                payload = String.valueOf( data );
+            }
+            else
+            {
+                payload = renderService.toJsonAsString( data );
+            }
+        }
+
+        audit.setData( payload );
+
+        jmsTemplate.send( destination, session -> session.createTextMessage( renderService.toJsonAsString( audit ) ) );
+    }
+
+    public void sendTopic( String topic, Audit audit )
+    {
+        send( new JmsTopic( topic ), audit );
+    }
+
+    public void sendTopic( String topic, Audit audit, Object data )
+    {
+        send( new JmsTopic( topic ), audit, data );
+    }
+
+    public void sendQueue( String queue, Audit audit )
+    {
+        send( new JmsQueue( queue ), audit );
+    }
+
+    public void sendQueue( String queue, Audit audit, Object data )
+    {
+        send( new JmsQueue( queue ), audit, data );
+    }
+}
```

Comment:
I think that for a client it should not be necessary to know if the audit message is handled by a topic or queue.
The AuditManager should be responsible for self-configuration and post the Audit message to the correct queue or topic.
Therefore, I'd simplify this interface with just an `audit` method which accept an `Audit` object and an option `Object` data. In fact, I'm not even sure the additional Object is really necessary, the Object could simply be a property of `Audit` - and we could add some logic there like `isPrimitive` `isIdentifiable` etc.